File: leetcode/search/1283. Find the Smallest Divisor Given a Threshold/main.cpp
int smallestDivisor(vector<int>& A, int threshold) {
  int left = 0;
  int right = 1e6;
  int sum = 0;

  while (left < right) {
    int m = left + ((right - left) / 2);
    sum = 0;

    for(int i : A) {
      sum += (i + m - 1) / m;
    }

    if (sum > threshold) {
      left = m + 1;
    } else {
      right = m;
    }
  }

  return left;
}

File: leetcode/search/1283. Find the Smallest Divisor Given a Threshold/js/main.js
// Time O(LogN)
// Space O(1)
const smallestDivisor = (nums, threshold) => {
  let left = 1;
  let right = 1e6;

  while (left < right) {
    let m = Math.floor((left + right) / 2);

    if (calc(m) > threshold) {
      left = m + 1;
    } else {
      right = m;
    }
  }

  return left;

  function calc(m) {
    let sum = 0;

    for (let i of nums) {
      sum += Math.floor((i + m - 1) / m);
    }

    return sum;
  }
};
